### PR TITLE
ranger: refine explain format, again

### LINKED ISSUE #6935: use <nil> to represent NULL is weird
let's say we have the following table:
```sql
drop table if exists t;
create table t(a char(20), index idx_a(a));
```

The following two queries result in the same explain result but they represent different DML queries:
```sql
desc select * from t where a = "<nil>";
desc select * from t where a is null;
```
```sql
TiDB(localhost:4000) > desc select * from t where a = "<nil>";
+-------------------+------+---------------------------------------------------------+-------+
| id                | task | operator info                                           | count |
+-------------------+------+---------------------------------------------------------+-------+
| IndexReader_9     | root | index:IndexScan_8                                       | 10.00 |
| └─IndexScan_8     | cop  | table:t, index:a, range:[<nil>,<nil>], keep order:false | 10.00 |
+-------------------+------+---------------------------------------------------------+-------+
2 rows in set (0.00 sec)

TiDB(localhost:4000) > desc select * from t where a is null;
+-------------------+------+---------------------------------------------------------+-------+
| id                | task | operator info                                           | count |
+-------------------+------+---------------------------------------------------------+-------+
| IndexReader_9     | root | index:IndexScan_8                                       | 10.00 |
| └─IndexScan_8     | cop  | table:t, index:a, range:[<nil>,<nil>], keep order:false | 10.00 |
+-------------------+------+---------------------------------------------------------+-------+
2 rows in set (0.00 sec)
```

Actually no mater what string we use to represent NULL, there is a chance that `a is null` and `a = xx` can result in the same range representation once the string we use to represent is equal to xx.

A way to solve this problem is to wrap strings with `"`, we can use the string `"NULL"` to represent NULL, which means we can distinguish NULL from other strings:
```sql
TiDB(localhost:4000) > desc select * from t where a = "NULL";
+-------------------+------+---------------------------------------------------------+-------+
| id                | task | operator info                                           | count |
+-------------------+------+---------------------------------------------------------+-------+
| IndexReader_9     | root | index:IndexScan_8                                       | 10.00 |
| └─IndexScan_8     | cop  | table:t, index:a, range:["NULL","NULL"], keep order:false | 10.00 |
+-------------------+------+---------------------------------------------------------+-------+
2 rows in set (0.00 sec)

TiDB(localhost:4000) > desc select * from t where a is null;
+-------------------+------+---------------------------------------------------------+-------+
| id                | task | operator info                                           | count |
+-------------------+------+---------------------------------------------------------+-------+
| IndexReader_9     | root | index:IndexScan_8                                       | 10.00 |
| └─IndexScan_8     | cop  | table:t, index:a, range:[NULL,NULL], keep order:false | 10.00 |
+-------------------+------+---------------------------------------------------------+-------+
2 rows in set (0.00 sec)
```

### LINKED ISSUE #7038: show whether we use the pseudo statistics for a table in the explain result
Take this as an example:
```sql
TiDB(localhost:4000) > drop table t;
Query OK, 0 rows affected (0.02 sec)

TiDB(localhost:4000) > create table t(a bigint, b bigint);
Query OK, 0 rows affected (0.14 sec)

TiDB(localhost:4000) > explain select * from t;
+-------------------+----------+------+----------------------------------------------+
| id                | count    | task | operator info                                |
+-------------------+----------+------+----------------------------------------------+
| TableReader_5     | 10000.00 | root | data:TableScan_4                             |
| └─TableScan_4     | 10000.00 | cop  | table:t, range:[-inf,+inf], keep order:false |
+-------------------+----------+------+----------------------------------------------+
2 rows in set (0.00 sec)
```
The problem is that we can hardly tell whether the row count estimated for `TableReader_5` is based on the pseudo statistics or the real statistics.

If the pseudo statistics is used, we can add "use pseudo stats" in the **"operator info"** field of `TableReader_5`.

### PR DESCRIPTION
**Pull request by zz-jason**
## What have you changed? (mandatory)

1. Refine the string format of ranges and constant expressions:
    - use `NULL` to represent NULL values instead of `<nil>`
    - wrap string with `"`

3. Show whether we use the pseudo statistics for a table or index in the explain result.
4. Refine the count and range information for inner table of IndexJoin

Before this PR:
```sql
TiDB(localhost:4000) > desc select * from t where a = "<nil>";
+-------------------+-------+------+---------------------------------------------------------+
| id                | count | task | operator info                                           |
+-------------------+-------+------+---------------------------------------------------------+
| IndexReader_9     | 10.00 | root | index:IndexScan_8                                       |
| └─IndexScan_8     | 10.00 | cop  | table:t, index:a, range:[<nil>,<nil>], keep order:false |
+-------------------+-------+------+---------------------------------------------------------+
2 rows in set (0.01 sec)

TiDB(localhost:4000) > desc select * from t where a = "NULL";
+-------------------+-------+------+-------------------------------------------------------+
| id                | count | task | operator info                                         |
+-------------------+-------+------+-------------------------------------------------------+
| IndexReader_9     | 10.00 | root | index:IndexScan_8                                     |
| └─IndexScan_8     | 10.00 | cop  | table:t, index:a, range:[NULL,NULL], keep order:false |
+-------------------+-------+------+-------------------------------------------------------+
2 rows in set (0.00 sec)

TiDB(localhost:4000) > desc select * from t where a is null;
+-------------------+-------+------+---------------------------------------------------------+
| id                | count | task | operator info                                           |
+-------------------+-------+------+---------------------------------------------------------+
| IndexReader_9     | 10.00 | root | index:IndexScan_8                                       |
| └─IndexScan_8     | 10.00 | cop  | table:t, index:a, range:[<nil>,<nil>], keep order:false |
+-------------------+-------+------+---------------------------------------------------------+
2 rows in set (0.00 sec)

TiDB(localhost:4000) > desc select /*+ TIDB_INLJ(t1) */ t2.a from t1 join t2 on t1.a=t2.a;
+--------------------------+----------+------+---------------------------------------------------------------------------+
| id                       | count    | task | operator info                                                             |
+--------------------------+----------+------+---------------------------------------------------------------------------+
| Projection_5             | 12500.00 | root | test.t2.a                                                                 |
| └─IndexJoin_9            | 12500.00 | root | inner join, inner:TableReader_8, outer key:test.t1.a, inner key:test.t2.a |
|   ├─TableReader_11       | 10000.00 | root | data:TableScan_10                                                         |
|   │ └─TableScan_10       | 10000.00 | cop  | table:t1, range:[-inf,+inf], keep order:false                             |
|   └─TableReader_8        | 10000.00 | root | data:TableScan_7                                                          |
|     └─TableScan_7        | 10000.00 | cop  | table:t2, range:[-inf,+inf], keep order:false                             |
+--------------------------+----------+------+---------------------------------------------------------------------------+
6 rows in set (0.00 sec)
```

After this PR:
```sql
TiDB(localhost:4000) > desc select * from t where a = "<nil>";
+-------------------+-------+------+---------------------------------------------------------------------------+
| id                | count | task | operator info                                                             |
+-------------------+-------+------+---------------------------------------------------------------------------+
| IndexReader_9     | 10.00 | root | index:IndexScan_8                                                         |
| └─IndexScan_8     | 10.00 | cop  | table:t, index:a, range:["<nil>","<nil>"], keep order:false, stats:pseudo |
+-------------------+-------+------+---------------------------------------------------------------------------+
2 rows in set (0.00 sec)

TiDB(localhost:4000) > desc select * from t where a = "NULL";
+-------------------+-------+------+-------------------------------------------------------------------------+
| id                | count | task | operator info                                                           |
+-------------------+-------+------+-------------------------------------------------------------------------+
| IndexReader_9     | 10.00 | root | index:IndexScan_8                                                       |
| └─IndexScan_8     | 10.00 | cop  | table:t, index:a, range:["NULL","NULL"], keep order:false, stats:pseudo |
+-------------------+-------+------+-------------------------------------------------------------------------+
2 rows in set (0.00 sec)

TiDB(localhost:4000) > desc select * from t where a is null;
+-------------------+-------+------+---------------------------------------------------------------------+
| id                | count | task | operator info                                                       |
+-------------------+-------+------+---------------------------------------------------------------------+
| IndexReader_9     | 10.00 | root | index:IndexScan_8                                                   |
| └─IndexScan_8     | 10.00 | cop  | table:t, index:a, range:[NULL,NULL], keep order:false, stats:pseudo |
+-------------------+-------+------+---------------------------------------------------------------------+
2 rows in set (0.00 sec)

TiDB(localhost:4000) > desc select /*+ TIDB_INLJ(t1) */ t2.a from t1 join t2 on t1.a=t2.a;
+--------------------------+----------+------+---------------------------------------------------------------------------+
| id                       | count    | task | operator info                                                             |
+--------------------------+----------+------+---------------------------------------------------------------------------+
| Projection_5             | 12500.00 | root | test.t2.a                                                                 |
| └─IndexJoin_9            | 12500.00 | root | inner join, inner:TableReader_8, outer key:test.t1.a, inner key:test.t2.a |
|   ├─TableReader_11       | 10000.00 | root | data:TableScan_10                                                         |
|   │ └─TableScan_10       | 10000.00 | cop  | table:t1, range:[-inf,+inf], keep order:false, stats:pseudo               |
|   └─TableReader_8        | 10.00    | root | data:TableScan_7                                                          |
|     └─TableScan_7        | 10.00    | cop  | table:t2, range: decided by [test.t1.a], keep order:false, stats:pseudo   |
+--------------------------+----------+------+---------------------------------------------------------------------------+
6 rows in set (0.01 sec)
```

## What is the type of the changes? (mandatory)

- Improvement (non-breaking change which is an improvement to an existing feature)

## How has this PR been tested? (mandatory)

- unit test
- explain test

## Refer to a related PR or issue link (optional)

fix #6935
fix #7038

### CODE REVIEW

**Comment by zz-jason** [comment]
@winoros PTAL

**Comment by shenli** [comment]
Please fix the CI.

File: util/ranger/types.go [hunk 1]
```diff
@@ -130,6 +130,9 @@ func formatDatum(d types.Datum, isLeftSide bool) string {
 		if d.GetUint64() == math.MaxUint64 && !isLeftSide {
 			return "+inf"
 		}
+	case types.KindString, types.KindBytes, types.KindMysqlEnum, types.KindMysqlSet,
+		types.KindMysqlJSON, types.KindBinaryLiteral, types.KindMysqlBit:
+		return fmt.Sprintf("\"%v\"", d.GetValue())
 	}
 	return fmt.Sprintf("%v", d.GetValue())
```
Comment (lamxTyler) [inline util/ranger/types.go]: Why not just wrap it here?

Comment (zz-jason) [reply util/ranger/types.go]: it's leaved for other types, for example decimal/double/datetime/timestamp, I think they'd better not be warped with `"`

**Comment by zz-jason** [comment]
@winoros @lamxTyler PTAL

File: util/ranger/ranger_test.go [hunk 2]
```diff
@@ -346,56 +346,56 @@ func (s *testRangerSuite) TestIndexRange(c *C) {
...
-			resultStr:   `[[\,\]]`,
+			resultStr:   "[[\"\\\",\"\\\"]]",
 		},
 		{
 			indexPos:    0,
 			exprStr:     `a LIKE "\\\\a%"`,
 			accessConds: `[like(test.t.a, \\a%, 92)]`,
 			filterConds: "[]",
-			resultStr:   `[[\a <nil>,\b <nil>)]`,
+			resultStr:   "[[\"\\a\" NULL,\"\\b\" NULL)]",
```
Comment (winoros) [inline util/ranger/ranger_test.go]: Why change the quotation?

**Comment by shenli** [comment]
@lamxTyler PTAL

File: util/ranger/ranger_test.go (hunk 2 above)
Comment (zz-jason) [reply util/ranger/ranger_test.go]: this is due to the modification of `formatDatum` in **util/ranger/types.go**

Comment (winoros) [reply util/ranger/ranger_test.go]: i mean use `, we don't need escape characters

Comment (zz-jason) [reply util/ranger/ranger_test.go]: Oh, the result was directly copied from the test output😂

Comment (zz-jason) [reply util/ranger/ranger_test.go]: The usage of ` and " is not uniformed now, there are many tests use " actually.

**Comment by zz-jason** [comment]
@winoros PTAL

File: plan/exhaust_physical_plans.go [hunk 3]
```diff
@@ -421,12 +421,94 @@ func (p *LogicalJoin) getIndexJoinByOuterIdx(prop *requiredProp, outerIdx int) [
...
+		rangeDecidedBy:   outerJoinKeys,
+	}.init(ds.ctx)
+	is.filterCondition = remainedConds
+
+	var rowCount float64
+	idxHist, ok := ds.statisticTable.Indices[idx.ID]
+	if ok && !ds.statisticTable.Pseudo {
+		rowCount = idxHist.AvgCountPerValue(ds.statisticTable.Count)
+	} else {
+		rowCount = ds.statisticTable.PseudoAvgCountPerCount()
```
Comment (coocood) [inline plan/exhaust_physical_plans.go]: The method name should be `PseudoAvgCountPerValue`?

Comment (zz-jason) [reply plan/exhaust_physical_plans.go]: yes, already changed.

**Comment by coocood** [comment]
LGTM

**Pull request merged**